File: app/src/main/java/ch/deletescape/lawnchair/iconpack/IconPackAdapter.java
package ch.deletescape.lawnchair.iconpack;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import ch.deletescape.lawnchair.R;

class IconPackAdapter extends RecyclerView.Adapter<IconPackAdapter.Holder> {

    private final Context mContext;
    private final List<EditIconActivity.IconPackInfo> mIconPacks;
    private Listener mListener;

    IconPackAdapter(Context context, List<EditIconActivity.IconPackInfo> iconPacks) {
        mContext = context;
        mIconPacks = iconPacks;
    }

    @Override
    public Holder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(mContext).inflate(R.layout.icon_pack_item, parent, false);
        return new Holder(itemView);
    }

    @Override
    public void onBindViewHolder(Holder holder, int position) {
        holder.bind(mIconPacks.get(position));
    }

    @Override
    public int getItemCount() {
        return mIconPacks.size();
    }

    private void onSelect(int position) {
        if (mListener == null) return;

        mListener.startPicker(mIconPacks.get(position));
    }

    public void setListener(Listener listener) {
        mListener = listener;
    }

    public class Holder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final ImageView mIcon;
        private final TextView mTitle;

        Holder(View itemView) {
            super(itemView);

            itemView.setOnClickListener(this);
            mIcon = itemView.findViewById(android.R.id.icon);
            mTitle = itemView.findViewById(android.R.id.title);
        }

        public void bind(EditIconActivity.IconPackInfo iconPackInfo) {
            mIcon.setImageDrawable(iconPackInfo.icon);
            mTitle.setText(iconPackInfo.label);
        }

        @Override
        public void onClick(View view) {
            onSelect(getAdapterPosition());
        }
    }

    interface Listener {

        void startPicker(EditIconActivity.IconPackInfo iconPackInfo);
    }
}

File: lawnchair/src/ch/deletescape/lawnchair/settings/adapters/IconPackAdapter.kt
package ch.deletescape.lawnchair.settings.adapters

import android.content.SharedPreferences
import android.graphics.drawable.Drawable
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import ch.deletescape.lawnchair.settings.views.IconPackListItem
import ch.deletescape.lawnchair.sharedprefs.LawnchairPreferences

class IconPackAdapter(private val dataSet: List<IconPackInfo>, prefs: SharedPreferences) :
    RecyclerView.Adapter<IconPackAdapter.ViewHolder>() {

    data class IconPackInfo(val name: String, val packageName: String, val icon: Drawable?)

    // TODO: Store package name of currently applied icon pack in settings database. Also see `IconPackSettingsFragment`.
    private var selectedIconPackPackageName: String = ""

    fun interface CheckListener {
        fun onItemChecked(id: String)
    }

    private val checkListener = CheckListener { id ->
        if (selectedIconPackPackageName != id) {
            selectedIconPackPackageName = id
            // TODO: Use `DiffUtil` instead of `notifyDataSetChanged` to fix animations.
            notifyDataSetChanged()
            prefs.edit().putString(LawnchairPreferences.ICON_PACK_PACKAGE, id).apply()
        }
    }

    class ViewHolder(private val view: IconPackListItem, private val checkListener: CheckListener) :
        RecyclerView.ViewHolder(view) {
        fun bind(iconPackInfo: IconPackInfo, selectedIconPackPackageName: String) {
            view.layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            view.iconPackName = iconPackInfo.name
            view.iconPackPackageName = iconPackInfo.packageName
            view.iconPackIcon = iconPackInfo.icon
            view.isChecked = selectedIconPackPackageName == view.iconPackPackageName
            itemView.setOnClickListener { checkListener.onItemChecked(view.iconPackPackageName) }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val iconPackListItem = IconPackListItem(parent.context)
        return ViewHolder(iconPackListItem, checkListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(dataSet[position], selectedIconPackPackageName)
    }

    override fun getItemCount(): Int = dataSet.size

}